In [5]:
import numpy as np

1.**load the data**

In [6]:
data = np.genfromtxt(
    "real_estate_dataset.csv" ,
    delimiter="," ,
    names=True,
    dtype=None ,
    encoding="utf-8"
)


In [8]:
#check the shape
data.shape

(500,)

In [9]:
#check the field names
data.dtype.names

('ID',
 'Square_Feet',
 'Num_Bedrooms',
 'Num_Bathrooms',
 'Num_Floors',
 'Year_Built',
 'Has_Garden',
 'Has_Pool',
 'Garage_Size',
 'Location_Score',
 'Distance_to_Center',
 'Price')

In [10]:
#Display the first 5 records
data[:5]

array([(1, 143.63502971, 1, 3, 3, 1967, 1, 1, 48, 8.2976312 ,  5.93573364, 602134.81674659),
       (2, 287.6785766 , 1, 2, 1, 1949, 0, 1, 37, 6.06146565, 10.8273922 , 591425.13538622),
       (3, 232.99848545, 1, 3, 2, 1923, 1, 0, 14, 2.91144248,  6.90459907, 464478.69687988),
       (4, 199.66462105, 5, 2, 2, 1918, 0, 0, 17, 2.07094918,  8.28401851, 583105.65599648),
       (5,  89.00466011, 4, 3, 3, 1999, 1, 0, 34, 1.52327786, 14.6482773 , 619879.14252279)],
      dtype=[('ID', '<i8'), ('Square_Feet', '<f8'), ('Num_Bedrooms', '<i8'), ('Num_Bathrooms', '<i8'), ('Num_Floors', '<i8'), ('Year_Built', '<i8'), ('Has_Garden', '<i8'), ('Has_Pool', '<i8'), ('Garage_Size', '<i8'), ('Location_Score', '<f8'), ('Distance_to_Center', '<f8'), ('Price', '<f8')])

2 **Analysing property prices**

In [13]:
#Extracting the prices
prices = data["Price"]

In [17]:
#Calculating the mean price
mean_price = np.mean(prices)
mean_price

np.float64(582209.6295285609)

In [18]:
#calculating the median
median_price = np.median
median_price

<function median at 0x000001D0C7BAF3B0>

In [19]:
#calculating the minimum and maximum price
min_price = np.min(prices)
max_price = np.max(prices)
min_price
max_price

np.float64(960678.274291193)

In [20]:
#calculating the percentiles
percentiles = np.percentile(prices,[25,50,75])
percentiles

array([503080.34413998, 574724.11334672, 665942.30127417])

In [22]:
#Extracting property sizes
property_size = data["Square_Feet"]

###Interpretation
Property prices vary across the dataset, with a difference between the minimum and maximum prices. The mean represents the average property price, while the median represents the middle price. The percentiles help show how property prices are distributed across the dataset.


3 **Analysing property sizes**

In [23]:
#Mean property size
mean_size = np.mean(property_size)
mean_size

np.float64(174.6404280585035)

In [24]:
#median
median_size = np.median(property_size)
median_size

np.float64(178.2909370891116)

In [26]:
#min and max
min_size = np.min(property_size)
max_size = np.max(property_size)
max_size
min_size

np.float64(51.265395961554674)

###Interpretation
Property sizes vary across the dataset. The mean and median show the typical property size, while the minimum and maximum show the range of property sizes. This indicates that the dataset contains properties of different sizes.

4 **calculating price per square foot**

In [27]:
#calculating for each property
price_per_sqft = prices / property_size

In [28]:
#mean
mean_price_per_sqft = np.mean(price_per_sqft)
mean_price_per_sqft

np.float64(4008.8199858500298)

In [29]:
#median
median_price_per_sqft = np.median(price_per_sqft)
median_price_per_sqft

np.float64(3361.191343306035)

In [30]:
#max and min
max_price_per_sqft = np.max(price_per_sqft)
min_price_per_sqft = np.min(price_per_sqft)
max_price_per_sqft
min_price_per_sqft

np.float64(1636.479625987531)

###Interpretation
The price per sqft shows how much buyers pay for each sqft of each property
The mean and median show the typical price per square foot, while the minimum and maximum show the range across the properties. Differences in price per square foot may reflect differences in property characteristics and location.


5 **Identify Expensive properties**

In [31]:
#finding the 75th percentile
top_25_threshold = np.percentile(prices, 75)
top_25_threshold

np.float64(665942.3012741683)

In [36]:
#creating the boolean mask
expensive_mask = prices >= top_25_threshold


In [38]:
#counting the expensive properties
num_expensive = np.sum(expensive_mask)
num_expensive

np.int64(125)

In [39]:
#Average price of expensive properties
expensive_prices = prices[expensive_mask]
average_expensive_price = np.mean(expensive_mask)
average_expensive_price

np.float64(0.25)

In [41]:
#overall average price
overall_average_price = np.mean(prices)
average_expensive_price, overall_average_price


(np.float64(0.25), np.float64(582209.6295285609))

### Interpretation

The top 25% of properties represent the most expensive properties in the dataset. Their average price is higher than the overall average property price, showing that these properties are substantially more expensive than a typical property in the dataset.

6 **Applying multiple conditions**

In [43]:
#Getting the location scores
location_score = data["Location_Score"]

In [44]:
#Creating the three conditions
condition_size = property_size > mean_size
condition_price = prices >= top_25_threshold
condition_location = location_score >=8


In [45]:
#combining all three
all_conditions = condition_size & condition_price & condition_location

In [46]:
#counting them
num_matching = np.sum(all_conditions)
num_matching

np.int64(17)

In [47]:
#display the matching properties
matching_properties = data[all_conditions]
matching_properties

array([(  8, 266.54403644, 5, 1, 3, 1973, 1, 1, 39, 9.37378438, 12.69278513, 875352.54518835),
       ( 63, 257.18437729, 4, 1, 3, 1941, 0, 0, 33, 9.94981369,  8.09137313, 695024.31070497),
       (113, 282.42441309, 3, 3, 2, 2015, 0, 0, 13, 9.01383097, 13.24967894, 787131.56969077),
       (127, 254.50369148, 3, 3, 1, 2005, 0, 0, 46, 9.92442289,  7.72423986, 743412.14388911),
       (151, 277.06647149, 5, 2, 3, 2019, 1, 1, 30, 9.03757572, 13.09418151, 960678.27429119),
       (201, 210.50791154, 5, 3, 2, 2000, 1, 1, 42, 8.18026749, 18.33206603, 885909.21527569),
       (273, 266.76807965, 5, 2, 1, 1977, 1, 0, 13, 9.52467117,  4.92572038, 789119.85805756),
       (314, 290.66210367, 4, 2, 1, 1996, 0, 0, 28, 9.90361782,  4.39413581, 789822.93141469),
       (323, 215.04934418, 5, 1, 3, 1992, 0, 0, 19, 9.78957507, 15.53606664, 747398.27978378),
       (383, 207.23571169, 5, 3, 3, 1967, 0, 1, 36, 8.07189319,  3.30604572, 845911.3698025 ),
       (384, 269.36800338, 5, 1, 1, 1933, 1, 0, 34

###Interpretation
Properties were filtered using three conditions: they had to be larger than the average property size, belong to the top 25% by price, and have a location score of at least 8. Boolean masking and the & operator were used to ensure that all three conditions were satisfied. The resulting records represent properties that meet all the specified characteristics.


7 **Compare the overage average**

In [48]:
overall_average_price = np.mean(prices)


In [51]:
#compare every price with the average
price_difference = prices-overall_average_price

In [52]:
#identify properties above the average
above_average = prices > overall_average_price


In [53]:
#identifying properties below average
below_average = prices < overall_average_price

In [54]:
np.sum(above_average), np.sum(below_average)

(np.int64(240), np.int64(260))

8 **finding the 10 most expensive properties**

In [ ]:
#sorting the price positions
sorted_indices = np.argsort(prices)

In [58]:
#get the 10 most expensive
top_10_indices = sorted_indices[-10:][::-1]


In [59]:
#Retrieving their information
top_10_properties = data[top_10_indices]
top_10_properties

array([(151, 277.06647149, 5, 2, 3, 2019, 1, 1, 30, 9.03757572, 13.09418151, 960678.27429119),
       ( 53, 284.87473539, 5, 2, 2, 2012, 1, 1, 21, 7.60854345, 13.94550838, 909199.56036649),
       (201, 210.50791154, 5, 3, 2, 2000, 1, 1, 42, 8.18026749, 18.33206603, 885909.21527569),
       (  8, 266.54403644, 5, 1, 3, 1973, 1, 1, 39, 9.37378438, 12.69278513, 875352.54518835),
       (474, 272.78577842, 5, 1, 2, 2017, 1, 0, 37, 6.32216598,  8.37164571, 867558.29676918),
       (500, 296.55268612, 4, 3, 1, 1988, 1, 1, 20, 7.89432194,  1.7797943 , 864299.5002175 ),
       (108, 238.88778464, 5, 3, 3, 1980, 1, 1, 43, 2.53821078, 15.17003762, 860013.25820393),
       (368, 276.33766049, 4, 3, 2, 2000, 0, 1, 43, 3.85613513, 10.27701469, 857487.18174448),
       (379, 253.61712065, 5, 3, 2, 1951, 1, 1, 32, 7.42824975, 14.83205239, 854374.5244548 ),
       ( 70, 296.72173415, 4, 1, 3, 2000, 1, 1, 18, 4.0667407 ,  4.15912719, 854304.12719063)],
      dtype=[('ID', '<i8'), ('Square_Feet', '<f8'

In [61]:
top_10_properties[[
    "ID",
    "Square_Feet",
    "Num_Bedrooms",
    "Location_Score",
    "Price"
]]

array([(151, 277.06647149, 5, 9.03757572, 960678.27429119),
       ( 53, 284.87473539, 5, 7.60854345, 909199.56036649),
       (201, 210.50791154, 5, 8.18026749, 885909.21527569),
       (  8, 266.54403644, 5, 9.37378438, 875352.54518835),
       (474, 272.78577842, 5, 6.32216598, 867558.29676918),
       (500, 296.55268612, 4, 7.89432194, 864299.5002175 ),
       (108, 238.88778464, 5, 2.53821078, 860013.25820393),
       (368, 276.33766049, 4, 3.85613513, 857487.18174448),
       (379, 253.61712065, 5, 7.42824975, 854374.5244548 ),
       ( 70, 296.72173415, 4, 4.0667407 , 854304.12719063)],
      dtype={'names': ['ID', 'Square_Feet', 'Num_Bedrooms', 'Location_Score', 'Price'], 'formats': ['<i8', '<f8', '<i8', '<f8', '<f8'], 'offsets': [0, 8, 16, 72, 88], 'itemsize': 96})

###interpretation
The property prices were sorted using np.argsort(), and the last 10 positions were selected to identify the most expensive properties. Fancy indexing was then used to retrieve their IDs, sizes, number of bedrooms, location scores, and prices. This makes it easier to examine the characteristics of the highest-priced properties.

9 **Final Analysis**

In [63]:
#Calculate the average for the top 10
top10_size_mean = np.mean(top_10_properties["Square_Feet"])
top10_bedrooms_mean = np.mean(top_10_properties["Num_Bedrooms"])
top10_location_mean = np.mean(top_10_properties["Location_Score"])

top10_size_mean, top10_bedrooms_mean, top10_location_mean

(np.float64(267.3895919331577), np.float64(4.7), np.float64(6.630599531469035))

In [65]:
#calculating the overall averages
overall_size_mean = np.mean(data["Square_Feet"])
overall_bedrooms_mean = np.mean(data["Num_Bedrooms"])
overall_location_mean = np.mean(data["Location_Score"])

overall_size_mean, overall_bedrooms_mean, overall_location_mean

(np.float64(174.6404280585035), np.float64(2.958), np.float64(5.1644099835214))

In [66]:
#comparing the top 10 with the overall dataset
print("Top 10 average size:", top10_size_mean)
print("Overall average size:", overall_size_mean)

print("Top 10 average bedrooms:", top10_bedrooms_mean)
print("Overall average bedrooms:", overall_bedrooms_mean)

print("Top 10 average location score:", top10_location_mean)
print("Overall average location score:", overall_location_mean)

Top 10 average size: 267.3895919331577
Overall average size: 174.6404280585035
Top 10 average bedrooms: 4.7
Overall average bedrooms: 2.958
Top 10 average location score: 6.630599531469035
Overall average location score: 5.1644099835214


***Business insight***

The 10 most expensive properties are generally larger than the average property, with an average size of about 267.39 square feet compared with 174.64 square feet overall. They also have more bedrooms on average, with 4.7 bedrooms compared with approximately 2.96 bedrooms overall. Their average location score is also higher at 6.63 compared with 5.16 for the overall dataset. These results indicate that larger property size, more bedrooms, and higher location scores are characteristics associated with higher property prices in this dataset.